File: programmers/algorithm_data_structure/stack_queue/주식가격.py
from collections import deque


def my_solution(prices):
    '''
    queue를 이용하여 정말로 정직하게 순서 따라가며 풀이
    그러나 속도가 빠르지 않다.
    '''
    answer = []
    queue = deque(prices)

    while queue:
        start_price = queue.popleft()
        answer.append(0)
        for price in queue:
            if start_price <= price:
                answer[-1] += 1

    return answer


def best_solution(p):
    '''
    queue 대신에 stack을 쓴 방법.
    코드가 이해하기 어려우나
    속도가 2배 이상 빠르다.

    stack: 주식 가격이 처음으로 떨어지는 지점을 아직 못찾은 가격의 index 모음    
    i for문을 돌며 'stack에 남은 것들이 i 번째에 처음으로 가격이 떨어지는가?'를 검사
    
    이때 queue를 쓰지 않고 stack을 써서 역으로 index를 검사하는 건
    stack 내 뒤쪽 것이 p[i]보다 가격이 같거나 작다면,
    그 앞의 것들은 i index에서 최초로 가격이 떨어질리 없기에 
    굳이 검사하지 않고 break로 시간을 줄일 수 있기 때문.
    '''

    ans = [0] * len(p)
    stack = [0]
    for i in range(1, len(p)):
        print('i:', i, ' stack:', stack, ' p[i]:', p[i],
         ' p[stack[-1]]:', p[stack[-1]])

        if p[i] < p[stack[-1]]:
            for j in stack[::-1]:
                print('i:', i, ' j:', j, ' p[i]:', p[i], ' p[j]:', p[j])
                if p[i] < p[j]:
                    ans[j] = i-j
                    stack.remove(j)
                else:
                    break
        stack.append(i)

        print('answer:', ans, '\n')

    for s in stack[:-1]:
        ans[s] = len(p) - s - 1
    return ans

File: programmers/algorithm_data_structure/stack_queue/주식가격.java
package algorithm_data_structure.stack_queue;

import java.util.*;

class 주식가격 {
    /*
     * beginIdxs : 아직 처리되지 않은(주식 가격 떨어지는 시점을 못찾은) 시점(index)
     * 
     * stack에 시점과 주식 가격을 같이 저장하는 방법을 처음에 썼지만
     * => idx만 알면 prices에서 조회하여 가격을 알 수 있어 idx만 저장하도록 수정
     */
    public int[] mySolution(int[] prices) {
        int[] answer = new int[prices.length];
        Stack<Integer> beginIdxs = new Stack<>();
        
        // 주식가격을 보면서 가격이 떨어지는 시점 찾기
        for (int currIdx = 0; currIdx < prices.length; currIdx++){
            while (!beginIdxs.isEmpty() && prices[beginIdxs.peek()] > prices[currIdx]){
                int popidx = beginIdxs.pop();

                answer[popidx] += currIdx - popidx;
            }
             
            beginIdxs.add(currIdx);                          
        }

        // 스택에 남은 값들 처리(가격이 끝까지 떨어지지 않음)
        while (! beginIdxs.isEmpty()){
            int popidx = beginIdxs.pop();
            answer[popidx] += ((prices.length -1) - popidx);
        }

        return answer;
    }
}